### PR TITLE
Part5

### PR DESCRIPTION
**Pull request by Naumych**

### CODE REVIEW

**Comment by svidlonix** [comment]
Don' t see ability to use almost all functionality in menu also please add ability to exit form program.

File: classes/cargo_train.rb [hunk 1]
```diff
@@ -5,4 +5,8 @@ class CargoTrain < Train
   def add_carriage
     carriages << CargoCarriage.new
   end
+
+  def to_s
+    "Cargo Train №#{number}"
```
Comment (svidlonix) [inline classes/cargo_train.rb]: This methods looks like the same https://github.com/Naumych/railsway/pull/3/files#diff-2f4e5957f6e2aad6148afbff9441d80eR9. Maybe move it to module ?

**Comment by svidlonix** [comment]
Please check if user try create new station with same name of station what have already existed, maybe good idea add verification and show message 'Station with this name has already existed.'

File: classes/train.rb [hunk 2]
```diff
@@ -1,18 +1,35 @@
 # frozen_string_literal: true
 
 require 'pry'
+require './modules/manufacture'
+require './modules/instances'
 
 # class Train
 class Train
-  attr_reader :route, :current_speed, :carriage_count, :current_station, :number, :type, :carriages
+  include Manufacture # почему в этом классе нужен require, а в других нет
```
Comment (svidlonix) [inline classes/train.rb]: Please don't use russian comments.

File: modules/manufacture.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+# Module which lets find out manufacturer's company: set and get manufacture_company
+module Manufacture
+  attr_reader :manufacture_company
```
[the remark this answers is not in the document]
Comment (Naumych) [reply modules/manufacture.rb]: If I don't announce InstanceMethods - I should use extend Instances::ClassMethods construction. And if I finally add InstanceMethods  - I should change old construction everywhere.

File: modules/instances.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      if instances[name].nil?
+        puts 'There are no object with this number/name'
+      else
+        puts "You asked me to find: #{instances[name]}"
+      end
+    end
+  end
+
+  # InstanceMethods
+  module InstanceMethods; end
```
[the remark this answers is not in the document]
Comment (Naumych) [reply modules/instances.rb]: If I don't announce InstanceMethods - I should use extend Instances::ClassMethods construction. And if I finally add InstanceMethods - I should change old construction everywhere.

File: classes/train.rb [hunk 5]
```diff
@@ -1,24 +1,30 @@
...
     @current_station = nil
     @route = nil
     @carriages = []
+    self.class.instance_hasher(@number, self)
+    self.class.superclass.instance_hasher(@number, self)
   end
 
   def gain_speed(speed)
-    raise 'Speed should be positive' if speed.negative?
+    raise 'Speed should be positive' if speed.to_f.negative?
```
Comment (svidlonix) [inline classes/train.rb]: Why do you need it ? Where is it should be rescued ?

File: classes/train.rb [hunk 6]
```diff
@@ -1,24 +1,30 @@
...
-  attr_reader :route, :current_speed, :carriage_count, :current_station, :number, :type, :carriages
+  include Manufacture
+  include Instances
+  attr_reader :route, :current_speed, :carriage_count, :current_station, :number, :carriages
 
   def initialize(number)
-    @number = number
+    @number = number.to_s
     @current_speed = 0
     @current_station_index = nil
```
Comment (svidlonix) [inline classes/train.rb]: Why do you need it ? It's 'nil' by default.

**Comment by Naumych** [comment]
> Don' t see ability to use almost all functionality in menu also please add ability to exit form program.

Fixed

**Comment by Naumych** [comment]
> Please check if user try create new station with same name of station what have already existed, maybe good idea add verification and show message 'Station with this name has already existed.'

Fixed. Add check to main methods "train/station/route creating"

**Comment by svidlonix** [comment]
@Naumych new comments

File: classes/train.rb (hunk 6 above)
Comment (Naumych) [reply classes/train.rb]: > Why do you need it ? It's 'nil' by default.

Fixed

File: classes/train.rb (hunk 5 above)
Comment (Naumych) [reply classes/train.rb]: > Why do you need it ? Where is it should be rescued ?

Fix with return

File: classes/cargo_train.rb (hunk 1 above)
Comment (Naumych) [reply classes/cargo_train.rb]: > This methods looks like the same https://github.com/Naumych/railsway/pull/3/files#diff-2f4e5957f6e2aad6148afbff9441d80eR9. Maybe move it to module ?

Temporary left

File: modules/manufacture.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Naumych) [reply modules/manufacture.rb]: > I think better use https://apidock.com/ruby/Module/attr_accessor here.

Fixed

**Pull request merged**